### PR TITLE
Docs for adding new distributions to Stan Math

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

This contains some fixes for the current getting started docs as well as documentation on how to write new distributions. It still needs some copy-editing that I'll have another pass over, but the structure should be all here. Tagging @spinkney, @adamhaber and @gregorp90 as they are involved in the discussion on adding the loglogistic distribution. If yinz have any time to read this over and let me know if it makes sense that would be great. There's definitely a few sections that need fleshed out more (and I can say I've never won a spelling bee in my life).

@t4c1 if you have time to look over this as well that'd be great. There's a few tricks I described which could use a once-over to make sure I said everything properly.

## Tests

You can generate the new docs by running `make doxygen`. They should appear inside of the docs folder and you can get to the pages by opening up `index.html` in your browser.

## Side Effects

Nope!

## Release notes

Improves documentation for the getting started guide and adds docs for contributing new distributions.

## Checklist

- [x] Math issue #2320 

- [x] Copyright holder: Steve Bronder

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+
+Each of the input arguments represent the inputs to the `Normal` function we wrote out above.
+The template parameters for univariate distributions are very general and they
+must work for all of Stan's scalar types
+`double`, `var`, and `fvar<T>` while accepting mixtures of scalars and vectors.
+The return of the function is the joint log probability accumulated over all
+of the inputs which is a scalar of the least upper bound of all the
+parameters scalar types. That's a lot of big words, but in essense means that
+if one of the inputs is a `var` and the others are double the return type needs to be
+a `double`. If the input signature contained `fvar<var>`, `var`, `double` then the
```
Comment (gregorp90) [inline doxygen/contributor_help_pages/adding_new_distributions.md:119]: Should this be "if one of the inputs is a `var` and the others are `double` the return type needs to be a `var`"?

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+function. One other nice thing about the matrixcalculus site is that it can generate latex
+which is nice for documentation.
+
+
+$$
+\begin{aligned}
+f = \text{Normal}(y|\mu,\sigma) &= -\frac{1}{2} \left(\frac{y-\mu}{\sigma}\right)^2 - \log(\sigma) - \frac{1}{2}\log(2\pi) \cr
+\frac{\partial f}{\partial y} &= -\frac{y-\mu}{\sigma^{2}} \cr
+\frac{\partial f}{\partial \mu} &= \frac{y-\mu}{\sigma^{2}} \cr
+\frac{\partial f}{\partial \sigma} &= -\frac{1}{\sigma} - \frac{(y-\mu)^{2}}{\sigma^{3}}
```
Comment (gregorp90) [inline doxygen/contributor_help_pages/adding_new_distributions.md:69]: The second sign should be +.

**Review by gregorp90 — COMMENTED** [review]
I added some comments. I might add some more, when I go through the rest of this documentation while working on loglogistic, if that's OK.

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+It's a little early, but once we get the `lpdf` function working with the above we will want to get out a pen and paper to simplify and find common subexpressions we only need to calculate once.
+For instance in the normal we can compute `y - mu` and `1/sigma`
+
+$$
+\begin{aligned}
+f(y|\mu,\sigma) = \text{Normal}(y|\mu,\sigma) &= -\frac{1}{2} \left(\frac{y-\mu}{\sigma}\right)^2 - \log(\sigma) - \frac{1}{2}\log(2\pi) \cr
+\frac{\partial f}{\partial y} &= -t_3 \cr
+\frac{\partial f}{\partial \mu} &= t_3 \cr
+\frac{\partial f}{\partial \sigma} &= t_3 \cdot t_0 - t_0 \cr
```
[the remark this answers is not in the document]
Comment (gregorp90) [reply doxygen/contributor_help_pages/adding_new_distributions.md:81]: Also, you can use the t_0 - t_3 already in line 78.

**Comment by SteveBronder** [comment]
> I added some comments. I might add some more, when I go through the rest of this documentation while working on loglogistic, if that's OK.

Much appreciated!!

> Maybe it is worth mentioning that if one of the inputs is a vector and others scalars, the operands_and_partials object will automatically sum the values for the scalar partials, if they are provided with a vector (as is the case in the current implementation of normal_lpdf, for example).

Yes that's def good to add. I'll fix these up today or tomorrow with the suggested changes. If anything else comes up def lmk!

**Comment by gregorp90** [comment]
> If anything else comes up def lmk!

Will do. Thanks for making this, I think it will be very helpful!

File: doxygen/contributor_help_pages/adding_new_distributions.md (hunk 1 above)
Comment (SteveBronder) [reply doxygen/contributor_help_pages/adding_new_distributions.md]: Lol yes extremely whoops on my part

File: doxygen/contributor_help_pages/adding_new_distributions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/contributor_help_pages/adding_new_distributions.md]: Actually I think how I wrong sigma's partial is wrong, I'm going to just put `\frac{t_{2}^2}{t_1}` until I think about it more.

**Comment by SteveBronder** [comment]
Alrighty I think this is ready for review! @gregorp90 would you mind skimming it over again to make sure anything here doesn't match up with what you did during the loglogistic PR? Or if there is anything missing that was important

**Comment by gregorp90** [comment]
@SteveBronder great, I will review it some time next week.

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+2. Write out the partial derivatives for each input
+3. Writing the function
+4. Testing the function
+
+
+### Get the Distribution Working In Stan
+
+We will use the normal distribution as an example and adding the lpdf
+function, though note for acceptance into Stan math a function must have
+it's respective `lpdf`, `lcdf`, `cdf`, `lccdf` and `rng` implimented.
```
Comment (gregorp90) [inline doxygen/contributor_help_pages/adding_new_distributions.md:19]: it's -> its. Same in some other cases throughout this text.
A typo in "implemented".

**Comment by gregorp90** [comment]
@SteveBronder I went through the docs, it looks good and I think it covers everything I did during the `loglogistic_lpdf` PR. The extended docs for getting_started.md also seem very helpful and some of the things I did while implementing functions in _rev_ are more clear now!

I added some minor comments to specific lines of code.

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+This is nice because now we can test this function against another implimentations
+to verify its correctness. At this point it is a good idea to post something on
+[discourse](https://discourse.mc-stan.org/) or file an
+[issue](https://github.com/stan-dev/math/issues) to let folks know you would like
+to add this distribution.
+
+
+### Writing out the Partials
+
+For an efficient implimentation of the distribution we want to calculate
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/contributor_help_pages/adding_new_distributions.md:52]: Bunch of typos, went and fixed these everywhere

File: doxygen/contributor_help_pages/adding_new_distributions.md [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+\frac{\partial f}{\partial \sigma} &= -\frac{1}{\sigma} + \frac{(y-\mu)^{2}}{\sigma^{3}}
+\end{aligned}
+$$
+
+It's a little early, but once we get the `lpdf` function working with the above we will want to get out a pen and paper to simplify and find common subexpressions we only need to calculate once.
+For instance in the normal we can compute `y - mu` and `1/sigma`
+
+$$
+\begin{aligned}
+f(y|\mu,\sigma) = \text{ln}\left(\text{Normal}(y|\mu,\sigma)\right) &= -\frac{1}{2} \left(\frac{y-\mu}{\sigma}\right)^2 - \log(\sigma) - \frac{1}{2}\log(2\pi) \cr
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply doxygen/contributor_help_pages/adding_new_distributions.md]: Umm we could add these though I don't hate leaving them as an exercise for the reader :-)

**Comment by SteveBronder** [comment]
@gregorp90 much appreciated! Yeah having another set of eyes on these are super helpful to sort out the bits and pieces. I fixed up the stuff you recommended and now I think this is good to go!

**Comment by SteveBronder** [comment]
@rok-cesnovar @t4c1 @andrjohns can one of yinz approve this?

**Review by rok-cesnovar — APPROVED** [review]

**Comment by rok-cesnovar** [comment]
Approving for @gregorp90.

**Pull request merged**